### DISCUSSION TITLE
How to create a target that isn't a SQL dialect?

### DISCUSSION THREAD

**Discussion by vanillajonathan** [PRQL/prql#2191] [Q&A]
It is quite obvious how to create a new target that is a SQL dialect, just edit `prql-compiler/src/sql/dialect.rs` and create a `struct` that implements the `DialectHandler` trait then provide some functions that override functions provided by the trait.

But how would one go about creating a target that isn't a SQL dialect?

**Comment by max-sixty** [comment]
We'd have to re-architect that part of the code to do that — probably we'd extract the translator code into a Trait, which would implement the translation from RQ to the target.

It's very possible, but it would be a sizable refactor...

**Reply by vanillajonathan** [reply]
This would make PRQL more flexible and open up more use cases and would be very useful as it would allow to translation support for Kusto, [Kinto](https://docs.kinto-storage.org/en/latest/api/1.x/filtering.html), Malloy, and maybe [SurrealDB](https://surrealdb.com/), [EdgeDB](https://www.edgedb.com/), [GraphQL](https://graphql.org/), XQuery, Cypher, etc.

**Reply by max-sixty** [reply]
Yes, the best targets are things that a) are widely used and b) people don't love writing to directly. 

So Kusto & Malloy & EdgeDB are good query languages in themselves and not _that_ widely used; I'm not sure there would be a lot of users.

That said, I agree there's potential for becoming more universal. It's probably some way off, but if anyone wants to write to a specific non-SQL target, let's reopen & discuss.

**Reply by snth** [reply]
I haven't thought much about the targets above but I have spent a bit of time about Pandas and Polars as possible targets. I'd quite like to be able to write PRQL and then have some codegen that produces Pandas code that I can copy into my Python script.

**Reply by max-sixty** [reply]
Yeah, ref https://github.com/PRQL/prql/issues/1099 — is there some functionality missing from the DuckDB approach?

**Reply by snth** [reply]
If you just care about the output and data, then I guess no and it's probably sufficient for my purposes. However for winning over existing Pandas users and more data scientists and ML engineers, I think demonstrably generating Pandas code would be more useful. I wouldn't make it a priority because of that but it's something I've considered a fair amount.

**Reply by max-sixty** [reply]
Yes, I def think it's possible. It would require someone who wanted to champion this — including refactoring the existing translator code to Traits. 

I also don't think it's high-priority for the project, but it would still be positive.

If anyone wants to do this, let's write a proposal for the refactor...

**Reply by eitsupi** [reply]
Perhaps we can start by implementing the conversion from the PL (json) output by the compiler, right?

**Reply by max-sixty** [reply]
Yes, from the RQ! (the PL is the first output, then RQ is second). This is the translator phase. 

It's a good point — it wouldn't actually _require_ a refactor, someone could implement a translator from RQ -> Pandas in lots of ways; including in any language.

(TBC, I think this offers almost no practical benefit over using DuckDB to target pandas, and I would vote against supporting it in the project, but as ever, open-minded :) )

**Reply by vanillajonathan** [reply]
More interestingly than Pandas, it opens up possibilities for other query languages,
